### PR TITLE
BUG: HDFStore failures on timezone-aware data

### LINKED ISSUE #20594: BUG: HDFStore failures on timezone-aware data
#### Code Sample, a copy-pastable example if possible

```python
In [1]: import pandas as pd
   ...: 
   ...: def check_roundtrip(obj):
   ...:     with pd.HDFStore('test.h5', 'w') as store:
   ...:         store['obj'] = obj
   ...:         retrieved = store['obj']
   ...:         return obj.equals(retrieved)
   ...: 
   ...: s = pd.Series([], dtype='datetime64[ns, UTC]')
   ...: t = pd.Series([0], dtype='datetime64[ns, UTC]')
   ...: df = pd.DataFrame({'A': s})

In [2]: print(check_roundtrip(s))
False

In [3]: print(check_roundtrip(t))
False

In [4]: print(check_roundtrip(df))

------------------------------------------------------------------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-57f1e9853bf6> in <module>()
----> 1 print(check_roundtrip(df))

<ipython-input-1-1d7fb84453e9> in check_roundtrip(obj)
      4     with pd.HDFStore('test.h5', 'w') as store:
      5         store['obj'] = obj
----> 6         retrieved = store['obj']
      7         return obj.equals(retrieved)
      8 

/home/ashieh/.local/lib/python2.7/site-packages/pandas/io/pytables.pyc in __getitem__(self, key)
    481 
    482     def __getitem__(self, key):
--> 483         return self.get(key)
    484 
    485     def __setitem__(self, key, value):

/home/ashieh/.local/lib/python2.7/site-packages/pandas/io/pytables.pyc in get(self, key)
    669         if group is None:
    670             raise KeyError('No object named %s in the file' % key)
--> 671         return self._read_group(group)
    672 
    673     def select(self, key, where=None, start=None, stop=None, columns=None,

/home/ashieh/.local/lib/python2.7/site-packages/pandas/io/pytables.pyc in _read_group(self, group, **kwargs)
   1347         s = self._create_storer(group)
   1348         s.infer_axes()
-> 1349         return s.read(**kwargs)
   1350 
   1351 

/home/ashieh/.local/lib/python2.7/site-packages/pandas/io/pytables.pyc in read(self, start, stop, **kwargs)
   2902             blk_items = self.read_index('block%d_items' % i)
   2903             values = self.read_array('block%d_values' % i,
-> 2904                                      start=_start, stop=_stop)
   2905             blk = make_block(values,
   2906                              placement=items.get_indexer(blk_items))

/home/ashieh/.local/lib/python2.7/site-packages/pandas/io/pytables.pyc in read_array(self, key, start, stop)
   2464             if shape is not None:
   2465                 # length 0 axis
-> 2466                 ret = np.empty(shape, dtype=dtype)
   2467             else:
   2468                 ret = node[start:stop]

TypeError: Invalid datetime unit in metadata string "[ns, UTC]"
```
#### Problem description

HDFStore fails to save empty/non-empty Series and empty DataFrames with timezone-aware data correctly. The issue is that the timezone information is not being saved correctly. For Series, there is no timezone handling at all. For DataFrames, the timezone handling is skipped when empty.

#### Expected Output

The checks should pass.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-1049-aws
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: 3.5.0
pip: 9.0.3
setuptools: 39.0.1
Cython: 0.28.1
numpy: 1.14.2
scipy: 1.0.1
pyarrow: 0.9.0
xarray: None
IPython: 5.6.0
sphinx: None
patsy: None
dateutil: 2.7.2
pytz: 2018.3
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: 0.8.0
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by adshieh**
- [x] closes #20594
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/test_pytables.py [hunk 1]
```diff
@@ -2793,10 +2793,20 @@ def test_empty_series_frame(self):
         self._check_roundtrip(df2, tm.assert_frame_equal)
 
     def test_empty_series(self):
-        for dtype in [np.int64, np.float64, np.object, 'm8[ns]', 'M8[ns]']:
+        for dtype in [np.int64, np.float64, np.object, 'm8[ns]', 'M8[ns]',
```
Comment (jschendel) [inline pandas/tests/io/test_pytables.py]: Can you use `@pytest.mark.parametrize` to parametrize this test over `dtype` instead of using a `for` loop?

Comment (adshieh) [reply pandas/tests/io/test_pytables.py]: Done!

File: doc/source/whatsnew/v0.23.0.txt [hunk 2]
```diff
@@ -1099,6 +1099,7 @@ I/O
 - Bug in :meth:`pandas.io.json.json_normalize` where subrecords are not properly normalized if any subrecords values are NoneType (:issue:`20030`)
 - Bug in ``usecols`` parameter in :func:`pandas.io.read_csv` and :func:`pandas.io.read_table` where error is not raised correctly when passing a string. (:issue:`20529`)
 - Bug in :func:`HDFStore.keys` when reading a file with a softlink causes exception (:issue:`20523`)
+- Bug in :class:`HDFStore` for Series and empty DataFrames with timezone-aware data (:issue:`20594`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: this is not very clear, is this for Series generally and empty DataFrames? this is *only* for a fixed store.

File: pandas/io/pytables.py [hunk 3]
```diff
@@ -2771,7 +2765,11 @@ def read(self, **kwargs):
     def write(self, obj, **kwargs):
         super(SeriesFixed, self).write(obj, **kwargs)
         self.write_index('index', obj.index)
-        self.write_array('values', obj.values)
+        if is_datetime64tz_dtype(obj.dtype):
```
Comment (jreback) [inline pandas/io/pytables.py]: this is reaching into the internal implementation way too much.

you can use ``obj.get_values()`` here

File: pandas/tests/io/test_pytables.py [hunk 4]
```diff
@@ -2792,10 +2792,22 @@ def test_empty_series_frame(self):
         self._check_roundtrip(df1, tm.assert_frame_equal)
         self._check_roundtrip(df2, tm.assert_frame_equal)
 
-    def test_empty_series(self):
-        for dtype in [np.int64, np.float64, np.object, 'm8[ns]', 'M8[ns]']:
-            s = Series(dtype=dtype)
-            self._check_roundtrip(s, tm.assert_series_equal)
+    @pytest.mark.parametrize('dtype', [
+        np.int64, np.float64, np.object, 'm8[ns]', 'M8[ns]',
+        'datetime64[ns, UTC]'
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: add US/Eastern as well

File: pandas/tests/io/test_pytables.py [hunk 5]
```diff
@@ -2792,10 +2792,22 @@ def test_empty_series_frame(self):
...
+    ])
+    def test_empty_series(self, dtype):
+        s = Series(dtype=dtype)
+        self._check_roundtrip(s, tm.assert_series_equal)
+
+    def test_series_timezone(self):
```
Comment (jreback) [inline pandas/tests/io/test_pytables.py]: parameterize these both with US/Eastern as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/pytables.py (hunk 3 above)
Comment (adshieh) [reply pandas/io/pytables.py]: The problem with `obj.get_values()` is that it returns an `array` with dtype `datetime64[ns]`, but we need a `DatetimeIndex` with dtype `datetime64[ns, tz]` to preserve the timezone information. I tried using `obj.dt.values`, which worked on `0.23.0`, but not on `master`.

Comment (jreback) [reply pandas/io/pytables.py]: so ``write_array`` already handles the objects corrctly, try just passing ``obj`` into that (you *may* need to add a case in write_array, but see where this gets you

**Comment by jreback** [comment]
can you update to last comments

**Comment by jreback** [comment]
can you rebase and respond to comments

**Pull request closed without merging**